core/src/main/java/org/apache/carbondata/core/util/CleanFilesUtil.java

```diff
@@ -0,0 +1,180 @@
+                + CarbonCommonConstants.FILE_SEPARATOR + CarbonTablePath.SEGMENT_PREFIX
+                + segmentNumber);
+            filesToDelete.add(FileFactory.getCarbonFile(file));
+          }
+        }
+        // After every file of that segment has been copied, need to delete those files.
+        LOGGER.info("Segment number: " + segmentNumber + "has been successfully copied to the" +
+            " trash folder");
+        try {
+          for (CarbonFile file : filesToDelete) {
+            FileFactory.deleteFile(file.getAbsolutePath());
+            // deleting empty  partition folders of partition table
+            SegmentFileStore.deleteEmptyPartitionFolders(FileFactory.getCarbonFile(new Path(file
+                .getAbsolutePath()).getParent().toString()));
+          }
+          filesToDelete.clear();
+        } catch (IOException e) {
+          LOGGER.error("Error while deleting the source data files. Please delete the files of" +
+              " segment: " + segmentNumber + " manually.", e);
+        }
+        // Delete the segment file too
+        FileFactory.deleteFile(CarbonTablePath.getSegmentFilePath(carbonTable.getTablePath(),
+            staleSegment));
+      }
+      staleSegments.clear();
+    }
+  }
+
+  /**
+   * This method will find all the stale segments by comparing the segment files in the
+   * metadata directory with the segments in the table status file. Any segment which has entry
+   * in the metadata folder and is not present in the table status file is considered as a
+   * stale segment. Only comparing from tablestatus file, not checking tablestatus.history file
+   */
+  private static List<String> getStaleSegments(CarbonTable carbonTable) {
+    // details contains segments in the tablestatus file, and all segments contains segments files.
+    // Segment number from those segment files is extracted and Stale segement file name is
+    // returned.
+    String metaDataLocation = carbonTable.getMetadataPath();
+    String segmentFilesLocation =
+        CarbonTablePath.getSegmentFilesLocation(carbonTable.getTablePath());
+    CarbonFile[] segmentFilesList = FileFactory.getCarbonFile(segmentFilesLocation).listFiles();
+    ArrayList<String> staleSegmentList = new ArrayList<>(segmentFilesList.length);
+    // there are no segments present in the Metadata folder. Can return here
+    if (segmentFilesList.length == 0) {
+      return staleSegmentList;
+    }
+    LoadMetadataDetails[] details = SegmentStatusManager.readLoadMetadata(metaDataLocation);
+    Set<String> loadNameSet = new HashSet<>(details.length);
+    Arrays.stream(details).map(loadMetadataDetails -> loadMetadataDetails.getLoadName())
+        .forEach(loadName -> loadNameSet.add(loadName));
+    for (CarbonFile segmentFile : segmentFilesList) {
+      if (!loadNameSet.contains(segmentFile.getName().split(CarbonCommonConstants
+          .UNDERSCORE)[0])) {
+        staleSegmentList.add(segmentFile.getName());
```

Comment:
taking the segment with the latest timestamp. For that, sorting the list of segment files and taking the one with the latest timestamp in case of same timestamp.